cover_agent/utils.py

```diff
@@ -0,0 +1,110 @@
+import logging
+import re
+import yaml
+
+from typing import List
+
+
+def load_yaml(response_text: str, keys_fix_yaml: List[str] = []) -> dict:
+    """
+    Load and parse YAML data from a given response text.
+
+    Parameters:
+    response_text (str): The response text containing YAML data.
+    keys_fix_yaml (List[str]): A list of keys to fix YAML formatting (default is an empty list).
+
+    Returns:
+    dict: The parsed YAML data.
+
+    If parsing the YAML data directly fails, it attempts to fix the YAML formatting using the 'try_fix_yaml' function.
+
+    Example:
+        load_yaml(response_text, keys_fix_yaml=['key1', 'key2'])
+
+    """
+    response_text = response_text.removeprefix("```yaml").rstrip("`")
+    try:
+        data = yaml.safe_load(response_text)
+    except Exception as e:
+        logging.error(f"Failed to parse AI prediction: {e}")
+        data = try_fix_yaml(response_text, keys_fix_yaml=keys_fix_yaml)
+    return data
+
+
+def try_fix_yaml(response_text: str, keys_fix_yaml: List[str] = []) -> dict:
+    """
+    Attempt to fix YAML formatting issues in the given response text.
+
+    Parameters:
+    response_text (str): The response text that may contain YAML data with formatting issues.
+    keys_fix_yaml (List[str]): A list of keys to fix YAML formatting (default is an empty list).
+
+    Returns:
+    dict: The parsed YAML data after attempting to fix formatting issues.
+
+    The function tries multiple fallback strategies to fix YAML formatting issues:
+    1. Tries to convert lines containing specific keys to multiline format.
+    2. Tries to extract YAML snippet enclosed between ```yaml``` tags.
+    3. Tries to remove leading and trailing curly brackets.
+    4. Tries to remove last lines iteratively to fix the formatting.
+
+    If none of the strategies succeed, an empty dictionary is returned.
+
+    Example:
+        try_fix_yaml(response_text, keys_fix_yaml=['key1', 'key2'])
+    """
+    response_text_lines = response_text.split("\n")
+
+    # first fallback - try to convert 'relevant line: ...' to relevant line: |-\n        ...'
+    response_text_lines_copy = response_text_lines.copy()
+    for i in range(0, len(response_text_lines_copy)):
+        for key in keys_fix_yaml:
+            if (
+                key in response_text_lines_copy[i]
+                and not "|-" in response_text_lines_copy[i]
+            ):
+                response_text_lines_copy[i] = response_text_lines_copy[i].replace(
+                    f"{key}", f"{key} |-\n        "
+                )
+    try:
+        data = yaml.safe_load("\n".join(response_text_lines_copy))
+        logging.info(f"Successfully parsed AI prediction after adding |-\n")
+        return data
+    except:
+        logging.info(f"Failed to parse AI prediction after adding |-\n")
+
+    # second fallback - try to extract only range from first ```yaml to ````
```

Comment:
A nice thing to have in a future PR would be also test for triple backticks surrounding the whole response. I see that happening a lot with mistral as it doesn't always follow the return format exactly.